Initial draft for triaging (#901)
nsidc/earthaccess: docs/contributing/triaging.md

### Issue Types

- **type: bug**: Use for issues that identify bugs causing incorrect or unexpected behavior.
- **type: duplicate**: Use this label for issues that are duplicates of existing ones.
- **type: enhancement**: Use this label for requests for new features or improvements to existing functionalities.
- **type: will not do**: Use this label for issues that won’t be addressed or fixed.

### Impact labels

These labels describe what portion of the project they affect:

- **impact: breaking**: Issues that break our public API.
- **impact: core**: Issues that affect the core Python codebase.
- **impact: dependencies**: Use this label for issues concerning dependencies.
- **impact: documentation**: Use this label for issues related to documentation.
- **impact: automation**: Use this label for issues related to the CI/CD pipeline or automation

### Needs labels

- **needs: decision**: We're struggling to decide what to do and the decision committee needs to help.
- **needs: feedback**:  Use this label for issues where feedback is requested from the team or our community.
- **needs: help**: Use this label for issues where additional help or contributions are needed.

> 

Reviewer: I'd like to add back the "good first issue" label, perhaps alone in a "Special labels" section? That is one that triagers should be especially aware of!

- Reporting bugs and errors.
- Tracking progress on specific tasks or projects.
- Requesting changes or improvements.

### When to Migrate

Use your best judgement when migrating between issues and discussions. Sometimes it makes more sense to open a new issue or discussion instead of migrating, for example when there are many things being discussed, but we want to create an issue or task out of just one of those things.

Migrate a discussion to an issue when:

- A specific task is identified.
- A bug or error is reported.
- A change or improvement is requested.

Migrate an issue to a discussion when:

- The issue is a nebulous idea that needs to be workshopped before it can be implemented.
- The issue is a general question or topic.
- The issue is not specific or actionable.

## Issue Triaging Workflow

``` mermaid

flowchart TD
  %%{init: {"flowchart": {"htmlLabels": false}} }%%
  classDef default font-size:32pt;
  start{"`Followed
  issue
  template?`"}
  start ==NO==> close1[Change
  to draft and ask to follow template]
  start == YES ==> dupe{Is duplicate?}
  dupe == YES ==> close2[Close and point to duplicate]
  dupe == NO ==> repro{Has proper reproduction?}

>   repro == NO ==> close3[Label: 'needs reproduction' bot will auto close if no update has been made in 3 days]

Reviewer: >bot will auto close if no update has been made in 3 days

I think this isn't true. We have one automation for autoclosing issues, but I believe it's broken: https://github.com/nsidc/earthaccess/blob/main/.github/workflows/issue-manager.yml

We should be careful about fixing it: #895 & https://jacobtomlinson.dev/posts/2024/most-stale-bots-are-anti-user-and-anti-contributor-but-they-dont-have-to-be/

Reviewer: Ah, that makes sense, thanks for the context!

Reviewer: I think getting @asteiker 's eyes on this is probably the best way forward. Will you be at next week's hack day? Maybe we can work together to finalize the flowchart.

Reviewer 2: @mfisher87 @Sherwin-14 Apologies for my radio silence on this recently! Yes I am happy to get back into this and work together at the next hackday. I'll review this in more detail in preparation.

Reviewer: Thanks, Amy!

  repro == YES ==> real{Is actually a bug?}
  real == NO ==> intended{Is the intended behaviour?}
  intended == YES ==> explain[Explain and close point to docs if needed]
  intended == NO ==> open[Keep open for discussion Remove 'pending triage' label]

>   real == YES ==> real2["1. Remove 'pending triage' label 2. Add related feature label if applicable (e.g. 'feat: ssr') 3. Add priority and meta labels (see below)"]

Reviewer: Renders as 
![image](https://github.com/user-attachments/assets/fbddb6d1-ad79-4c7c-bf85-892ae8cf9f12)

  real2 ==> unusable{Does the bug make earthaccess unusable?}
  unusable == YES ==> maj{Does the bug affect the majority of earthaccess users?}
  maj == YES ==> p5[p5: urgent]
  maj == NO ==> p4[p4: important]
  unusable == NO ==> workarounds{Are there workarounds for the bug?}
  workarounds == NO ==> p3[p3: minor bug]

>   workarounds == YES ==> p2[p2: edge case has workaround]

Reviewer: We don't have priority levels yet to my knowledge. @asteiker what do you think?

Reviewer: I think that sounds reasonable!

- **type: bug**: Use for issues that identify bugs causing incorrect or unexpected behavior.
- **type: duplicate**: Use this label for issues that are duplicates of existing ones.
- **type: enhancement**: Use this label for requests for new features or improvements to existing functionalities.
- **type: will not do**: Use this label for issues that won’t be addressed or fixed.

### Impact labels

These labels describe what portion of the project they affect:

- **impact: breaking**: Issues that break our public API.
- **impact: core**: Issues that affect the core Python codebase.
- **impact: dependencies**: Use this label for issues concerning dependencies.
- **impact: documentation**: Use this label for issues related to documentation.
- **impact: automation**: Use this label for issues related to the CI/CD pipeline or automation

### Needs labels

- **needs: decision**: We're struggling to decide what to do and the decision committee needs to help.
- **needs: feedback**:  Use this label for issues where feedback is requested from the team or our community.
- **needs: help**: Use this label for issues where additional help or contributions are needed.

### Special labels

> - **good-first-issue**: Use this label for issues that are suitable for new contributors. These issues are designed to be approachable and not overly complex, making them an ideal starting point for those looking to contribute to the project.

Reviewer: ```suggestion
- **good first issue**: Use this label for issues that are suitable for new contributors. These issues are designed to be approachable and not overly complex, making them an ideal starting point for those looking to contribute to the project.
```

- **library**: Use this label for issues related to core libraries or preparations for future plugin impacts.

### Requests and Feedback

- **enhancement**: Use this label for requests for new features or improvements to existing functionalities.

- **feedback**:  Use this label for issues where feedback is requested from the team or our community.

- **help**: Use this label for issues where additional help or contributions are needed.

### Special Labels

- **automation**: Use this label for issues related to the CI/CD pipeline or automation

- **duplicate**: Use this label for issues that are duplicates of existing ones.

- **good first issue** : Use this label to highlight issues good for newcomers who want to contribute to Earthaccess.

- **hackathon**: Use this label for issues to be discussed during hackdays.

- **wontfix**: Use this label for issues that won’t be addressed or fixed.

## Linking Labels in GitHub Markdown

When referencing a label in a GitHub issue or discussion, it will be  useful to link to the label page to provide additional context and help other members to quickly understand the issue's category.

### Syntax

> To link to a label in GitHub Markdown, use the following syntax:
> 
> ```
> [label name](https://github.com/username/repository/labels/label-name)
> ```

Reviewer: ````suggestion
To link to a label in GitHub Markdown, copy-and-paste the URL to the label by right-clicking [any label](https://github.com/nsidc/earthaccess/labels) and selecting "Copy Link". Then, paste that label in a GitHub issue, PR, discussion, or Markdown document:

```
https://github.com/username/repository/labels/label-name
```
````

### Example

For example, to link to the "hackathon" label in the nsidc/earthaccess repository, you would use the following syntax:

```

> [hackathon](https://github.com/nsidc/earthaccess/labels/hackathon)

Reviewer: Should actually look like https://github.com/nsidc/earthaccess/labels/good%20first%20issue - GitHub automatically applies the text when you paste the URL :)

```
Should actually look like https://github.com/nsidc/earthaccess/labels/good%20first%20issue
```

> [hackathon](https://github.com/nsidc/earthaccess/labels/hackathon)

Reviewer: After our example, let's include an image that shows what it should look like when done correctly!

```

> !!! tip
>     When linking to labels, it is a good practice to use the label name as the link text, as shown in the example above. This makes it clear to others what label is being referenced, and helps to avoid confusion.

Reviewer: ```suggestion
```

This is a great tip, but if the label syntax is correct, the user shouldn't have to worry about that because GitHub handles the magic :)

## Close Issue as Not Planned

> This section will cover the guidelines for when to use the "Close Issue as Not Planned" label, and how to handle issues that are not planned or feasible.

Reviewer: ```suggestion
This section will cover the guidelines for when to use the "Close Issue as Not Planned" feature, and how to handle issues that are not planned or feasible.
```

> ### When to Use This Label?

Reviewer: ```suggestion
### When to "Close as not planned"?
```

> Use the "Close Issue as Not Planned" label when:

Reviewer: ```suggestion
Close issues as "not planned" when:
```

- An issue is not aligned with the project's goals or priorities.
- An issue is not feasible to be addressed due to technical or resource constraints.
- An issue is a duplicate of an existing issue that has already been addressed.

When closing an issue as not planned, keep the following best practices in mind:

- Provide a clear explanation as to why the issue is not planned or feasible.
- Offer alternative solutions or workarounds, if possible.

> - Link to relevant documentation or resources, if applicable.

Reviewer: :100: Amazing callout!

## Discussions vs Issues

This section would cover the guidelines for when to use discussions versus issues, and how to migrate between them.

###  What are Discussions?

Discussions are used for:

- Brainstorming and idea generation.

> - Feature requests and feedback.

Reviewer: ```suggestion
- Project feedback.
```

I wanted to differentiate this from "Requesting changes or improvements" in the "What are issues?" section :)

- General questions and topics.

###  What are Issues?

Issues are used for:

- Reporting bugs and errors.
- Tracking progress on specific tasks or projects.
- Requesting changes or improvements.

> ### When to Migrate

Reviewer: ```suggestion
### When to Migrate

Use your best judgement when migrating between issues and discussions. Sometimes it makes more sense to open a new issue or discussion instead of migrating, for example when there are many things being discussed, but we want to create an issue or task out of just one of those things.
```

Migrate a discussion to an issue when:

- A specific task is identified.
- A bug or error is reported.
- A change or improvement is requested.

Migrate an issue to a discussion when:

> - The issue is a feature request or idea.

Reviewer: ```suggestion
- The issue is a nebulous idea that needs to be workshopped before it can be implemented.
```
